### PR TITLE
Quantizer EXE, OPQ Support

### PR DESCRIPTION
**Pull request by PhilipBAdams**
* Add PQ Quantizer Training support

* Add OPQ support for index build/index search

### CODE REVIEW

File: AnnService/src/Core/VectorIndex.cpp [hunk 1]
```diff
@@ -368,6 +368,9 @@ VectorIndex::SaveIndexToFile(const std::string& p_file, IAbortOperation* p_abort
     IOBINARY(fp, WriteBinary, sizeof(configSize), (char*)&configSize);
     if ((ret = SaveIndexConfig(fp)) != ErrorCode::Success) return ret;
     configSize = fp->TellP() - sizeof(configSize);
+    if (ErrorCode::Success == ret && m_pQuantizer) {
```
Comment (MaggieQi) [inline AnnService/src/Core/VectorIndex.cpp:371]: Why we need to change the load quantizer position? If you want to change, then you need to change all of the save index and load index as well as the buffersizes and filenames order you get from the index.

File: AnnService/src/IndexBuilder/main.cpp [hunk 2]
```diff
@@ -101,9 +105,14 @@ int main(int argc, char* argv[])
         code = indexBuilder->BuildIndex(vectorReader->GetVectorSet(), vectorReader->GetMetadataSet(), options->m_metaMapping, options->m_normalized);
     }
     else {
+        indexBuilder->SetQuantizerFileName(options->m_quantizerFile.substr(options->m_quantizerFile.find_last_of("/\\") + 1));
         code = indexBuilder->BuildIndex(options->m_normalized);    
     }
-    if (code == ErrorCode::Success) indexBuilder->SaveIndex(options->m_outputFolder);
+    if (code == ErrorCode::Success)
+    {
+
```
Comment (MaggieQi) [inline AnnService/src/IndexBuilder/main.cpp:113]: remove the empty line?

**Review by MaggieQi — CHANGES_REQUESTED** [review]

File: AnnService/inc/Core/SPANN/ExtraFullGraphSearcher.h [hunk 3]
```diff
@@ -106,9 +106,9 @@ namespace SPTAG
                     auto curIndexFile = f_createAsyncIO();
                     if (curIndexFile == nullptr || !curIndexFile->Initialize(curFile.c_str(), std::ios::binary | std::ios::in, 
 #ifdef BATCH_READ
-                        p_opt.m_searchInternalResultNum, 2, 2, p_opt.m_iSSDNumberOfThreads
+                        max(p_opt.m_searchInternalResultNum*m_vectorInfoSize, 1 << 12), 2, 2, p_opt.m_iSSDNumberOfThreads
```
[the remark this answers is not in the document]
Comment (PhilipBAdams) [reply AnnService/inc/Core/SPANN/ExtraFullGraphSearcher.h:109]: Sorry, can you share the PR? I had some trouble finding it

File: AnnService/src/Core/VectorIndex.cpp (hunk 1 above)
Comment (PhilipBAdams) [reply AnnService/src/Core/VectorIndex.cpp:371]: Before the local-to-index quantizer change, it was needed to load the quantizer first so the right distance function would be picked. Now, it is not needed - I think it is changed in all the right places, but I can revert it.

Comment (MaggieQi) [reply AnnService/src/Core/VectorIndex.cpp:371]: It is better to keep original load order : )

Comment (PhilipBAdams) [reply AnnService/src/Core/VectorIndex.cpp]: returned the load order in most recent commit

**Review by MaggieQi — APPROVED** [review]

**Pull request merged**